src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -2132,6 +2132,90 @@ void C2_MacroAssembler::evpcmp(BasicType typ, KRegister kdmask, KRegister ksmask
   }
 }
 
+void C2_MacroAssembler::vpcmpu(BasicType typ, XMMRegister dst, XMMRegister src1, XMMRegister src2, ComparisonPredicate comparison,
+                            int vlen_in_bytes, XMMRegister vtmp1, XMMRegister vtmp2, Register scratch) {
+  int vlen_enc = vector_length_encoding(vlen_in_bytes*2);
+  switch (typ) {
+  case T_BYTE:
+    vpmovzxbw(vtmp1, src1, vlen_enc);
+    vpmovzxbw(vtmp2, src2, vlen_enc);
+    vpcmpCCW(dst, vtmp1, vtmp2, comparison, Assembler::W, vlen_enc, scratch);
+    vpand(dst, dst, ExternalAddress(StubRoutines::x86::vector_short_to_byte_mask()), vlen_enc, scratch);
+    vpackuswb(dst, dst, dst, vlen_enc);
+    break;
+  case T_SHORT:
+    vpmovzxwd(vtmp1, src1, vlen_enc);
+    vpmovzxwd(vtmp2, src2, vlen_enc);
+    vpcmpCCW(dst, vtmp1, vtmp2, comparison, Assembler::D, vlen_enc, scratch);
+    vpand(dst, dst, ExternalAddress(StubRoutines::x86::vector_int_to_short_mask()), vlen_enc, scratch);
+    vpackusdw(dst, dst, dst, vlen_enc);
+    break;
+  case T_INT:
+    vpmovzxdq(vtmp1, src1, vlen_enc);
+    vpmovzxdq(vtmp2, src2, vlen_enc);
+    vpcmpCCW(dst, vtmp1, vtmp2, comparison, Assembler::Q, vlen_enc, scratch);
+    vpermilps(dst, dst, 8, vlen_enc);
+    break;
+  default:
+    assert(false, "Should not reach here");
+  }
+  if (vlen_in_bytes == 16) {
+    vpermpd(dst, dst, 0x8, vlen_enc);
+  }
+}
+
+void C2_MacroAssembler::vpcmpu32(BasicType typ, XMMRegister dst, XMMRegister src1, XMMRegister src2, ComparisonPredicate comparison, int vlen_in_bytes,
+                              XMMRegister vtmp1, XMMRegister vtmp2, XMMRegister vtmp3, Register scratch) {
+  int vlen_enc = vector_length_encoding(vlen_in_bytes);
+  switch (typ) {
+  case T_BYTE:
+    vpmovzxbw(vtmp1, src1, vlen_enc);
+    vpmovzxbw(vtmp2, src2, vlen_enc);
+    vpcmpCCW(dst, vtmp1, vtmp2, comparison, Assembler::W, vlen_enc, scratch);
+    vpand(dst, dst, ExternalAddress(StubRoutines::x86::vector_short_to_byte_mask()), vlen_enc, scratch);
+    vextracti128(vtmp1, src1, 1);
+    vextracti128(vtmp2, src2, 1);
+    vpmovzxbw(vtmp1, vtmp1, vlen_enc);
+    vpmovzxbw(vtmp2, vtmp2, vlen_enc);
+    vpcmpCCW(vtmp3, vtmp1, vtmp2, comparison, Assembler::W, vlen_enc, scratch);
+    vpand(vtmp3, vtmp3, ExternalAddress(StubRoutines::x86::vector_short_to_byte_mask()), vlen_enc, scratch);
+    vpackuswb(dst, dst, vtmp3, vlen_enc);
+    vpermpd(dst, dst, 0xd8, vlen_enc);
```

Comment:
since comparison is performed at lane level (x86 definition 128 bits) and later on we are packing the results of two lanes I am not sure about the need for last permute instruction.

Reply:
If you look at the packuswb, the packing is done per 128 bit alternating from src1 and src2.
i.e. 
0-127 bits from src1 -> 0-63 bits in dst
0-127 bits from src2 -> 64-127 bits in dst
128-255 bits from src1 -> 128-191 bits in dst
128-255 bits from src2 -> 192-255 bits in dst
The src1 and src2 are mish-mashed in dst and need to be put in their proper place by permpd.

Reply:
Yes that's corrects. thanks!